src/java.base/share/classes/jdk/internal/reflect/ReflectionFactory.java

```diff
@@ -690,6 +691,8 @@ private static void checkInitted() {
         disableSerialConstructorChecks =
             "true".equals(props.getProperty("jdk.disableSerialConstructorChecks"));
 
+        // ensure previous fields are visible before initted is
+        Unsafe.getUnsafe().storeStoreFence();
```

Comment:
Ensuring ordering on the writer side, without also ensuring ordering on the reader side, doesn't solve an ordering problem. Just make `initted` volatile and this should be safe from a Java Memory Model perspective.

Reply:
On a side note, would a store load fence (like `VarHandle.storeFence`) after the write be a valid alternative, so the field itself can be get plain?

Reply:
I don't see such a "fence" in the VarHandle API. ?? The `VarHandle.storeStoreFence()` is no different in memory ordering properties to the `Unsafe.storeStoreFence()`.